src/LetsEncrypt/Internal/CertificateFactory.cs

```diff
@@ -183,50 +186,44 @@ private async Task ValidateDomainOwnershipAsync(string domainName, CancellationT
             throw new TimeoutException("Timed out waiting for domain ownership validation.");
         }
 
-        private Exception InvalidAuthorizationError(string domainName, AcmeResult<AuthorizationEntity> authorization)
+        private Exception InvalidAuthorizationError(Authorization authorization)
         {
             var reason = "unknown";
+            var domainName = authorization.Identifier.Value;
             try
             {
-                var errorStub = new { error = new { type = "", detail = "", status = -1 } };
-                var data = JsonConvert.DeserializeAnonymousType(authorization.Json, errorStub);
-                reason = $"{data.error.type}: {data.error.detail}, Code = {data.error.status}";
+                var error = authorization.Challenges.First(a => a.Error != null).Error;
+                reason = $"{error.Type}: {error.Detail}, Code = {error.Status}";
             }
             catch
             {
-                _logger.LogTrace("Could not determine reason why validation failed. Response: {resp}", authorization.Json);
+                _logger.LogTrace("Could not determine reason why validation failed. Response: {resp}", authorization);
             }
 
             _logger.LogError("Failed to validate ownership of domainName '{domainName}'. Reason: {reason}", domainName, reason);
 
             return new InvalidOperationException($"Failed to validate ownership of domainName '{domainName}'");
         }
 
-        private async Task<X509Certificate2> CompleteCertificateRequestAsync(CancellationToken cancellationToken)
+        private async Task<X509Certificate2> CompleteCertificateRequestAsync(IOrderContext order, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
-            var distinguishedName = "CN=" + _options.Value.DomainNames[0];
-            _logger.LogDebug("Creating cert for {distinguishedName}", distinguishedName);
+            var commonName = _options.Value.DomainNames[0];
+            _logger.LogDebug("Creating cert for {commonName}", commonName);
 
-            var csb = new CertificationRequestBuilder();
-            csb.AddName(distinguishedName);
-            foreach (var name in _options.Value.DomainNames.Skip(1))
+            var csrInfo = new CsrInfo
             {
-                csb.SubjectAlternativeNames.Add(name);
```

Comment:
Where are SAN extensions added to the CSR? That's crucial to making multi-domain certificates work

Reply:
Well, multi-domain certificates just work, they somehow simplified it, the list of domain names is provided at the beginning, line 66. And then at the end the certificate has them all :-)